moveit_ros/hybrid_planning/local_planner/local_constraint_solver_plugins/src/servo_solver.cpp

```diff
@@ -0,0 +1,174 @@
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author: Sebastian Jahr, Adam Pettinger
+ */
+
+#include <control_msgs/msg/joint_jog.hpp>
+#include <geometry_msgs/msg/twist_stamped.hpp>
+#include <moveit/local_constraint_solver_plugins/servo_solver.hpp>
+#include <moveit/planning_scene/planning_scene.h>
+#include <moveit/robot_state/conversions.h>
+#include <moveit/utils/logger.hpp>
+
+namespace moveit::hybrid_planning
+{
+namespace
+{
+rclcpp::Logger getLogger()
+{
+  return moveit::getLogger("servo_solver");
+}
+}  // namespace
+
+bool ServoSolver::initialize(const rclcpp::Node::SharedPtr& node,
+                             const planning_scene_monitor::PlanningSceneMonitorPtr& planning_scene_monitor,
+                             const std::string& /*group_name*/)
+{
+  planning_scene_monitor_ = planning_scene_monitor;
+  node_ = node;
+
+  const std::shared_ptr<const servo_solver_parameters::ParamListener> solver_param_listener =
+      std::make_shared<const servo_solver_parameters::ParamListener>(node, "");
+  solver_parameters_ = solver_param_listener->get_params();
+
+  // Get Servo Parameters
+  const std::string param_namespace = "moveit_servo";
+  const std::shared_ptr<const servo::ParamListener> servo_param_listener =
+      std::make_shared<const servo::ParamListener>(node, param_namespace);
+  servo_parameters_ = servo_param_listener->get_params();
+
+  // Create Servo and start it
+  servo_ = std::make_unique<moveit_servo::Servo>(node_, servo_param_listener, planning_scene_monitor_);
+
+  // Use for debugging
+  // twist_cmd_pub_ = node_->create_publisher<geometry_msgs::msg::TwistStamped>("~/delta_twist_cmds", 10);
+  return true;
+}
+
+bool ServoSolver::reset()
+{
+  RCLCPP_INFO(getLogger(), "Reset Servo Solver");
+  joint_cmd_rolling_window_.clear();
+  return true;
+};
+
+moveit_msgs::action::LocalPlanner::Feedback
+ServoSolver::solve(const robot_trajectory::RobotTrajectory& local_trajectory,
+                   const std::shared_ptr<const moveit_msgs::action::LocalPlanner::Goal> /*local_goal*/,
+                   trajectory_msgs::msg::JointTrajectory& local_solution)
+{
+  // Create Feedback
+  moveit_msgs::action::LocalPlanner::Feedback feedback_result;
+
+  // Transform next robot trajectory waypoint into JointJog message
+  moveit_msgs::msg::RobotTrajectory robot_command;
+  local_trajectory.getRobotTrajectoryMsg(robot_command);
+
+  if (robot_command.joint_trajectory.points.empty())
+  {
+    feedback_result.feedback = std::string("Reference trajectory does not contain any points");
+    return feedback_result;
+  }
+
+  // Get current state
+  const auto current_state = planning_scene_monitor_->getStateMonitor()->getCurrentState();
```

Comment:
> If you are certain that the current state is actually current, then I would add it into the command queue as the first state.

Reply:
I think it might not be the current state anymore since the robot is moving while local planning takes place. What do you think is the tolerance there?